ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -21,81 +28,82 @@ pub struct ProofVerifier<F: PrimeField> {
 /// `https://eprint.iacr.org/2023/909.pdf`
 ///
 #[allow(non_camel_case_types)]
-impl<F> ProofVerifier<F>
+impl<F, λ> ProofVerifier<F, λ>
 where
     F: PrimeField,
+    λ: ArrayLength,
 {
-    pub fn new(u_or_v: Vec<F>, out_share: F) -> Self {
+    pub fn new(u_or_v: Vec<GenericArray<F, λ>>, out_share: F) -> Self {
         Self { u_or_v, out_share }
     }
 
-    pub fn verify_proof<λ>(
-        &self,
+    pub fn verify_proof<J, B>(
+        mut u_or_v_iterator: J,
+        out_share: F,
         zkp: &ZeroKnowledgeProof<F, TwoNMinusOne<λ>>,
         r: F,
-    ) -> (F, ProofVerifier<F>)
+    ) -> (F, Self)
     where
-        λ: ArrayLength + Add + Sub<U1>,
+        λ: Add + Sub<U1>,
         <λ as Add>::Output: Sub<U1>,
         <<λ as Add>::Output as Sub<U1>>::Output: ArrayLength,
         <λ as Sub<U1>>::Output: ArrayLength,
+        J: Iterator<Item = B>,
+        B: Borrow<GenericArray<F, λ>>,
     {
-        debug_assert_eq!(self.u_or_v.len() % λ::USIZE, 0); // We should pad with zeroes eventually
-
-        let s = self.u_or_v.len() / λ::USIZE;
-
-        assert!(
-            s > 1,
-            "When the output is this small, you should call `verify_final_proof`"
-        );
-
         let denominator_g = CanonicalLagrangeDenominator::<F, TwoNMinusOne<λ>>::new();
         let lagrange_table_g = LagrangeTable::<F, TwoNMinusOne<λ>, U1>::new(&denominator_g, &r);
         let g_r_share = lagrange_table_g.eval(&zkp.g)[0];
         let sum_share = (0..λ::USIZE).fold(F::ZERO, |acc, i| acc + zkp.g[i]);
 
         // Reveal `b_share` to one another to reconstruct `b` and check if `b = 0`. If the check doesn't pass, abort.
-        let b_share = sum_share - self.out_share;
+        let b_share = sum_share - out_share;
 
         let denominator_p_or_q = CanonicalLagrangeDenominator::<F, λ>::new();
         let lagrange_table_p_or_q_r = LagrangeTable::<F, λ, U1>::new(&denominator_p_or_q, &r);
-        let p_or_q_r = (0..s)
-            .map(|i| {
-                let start = i * λ::USIZE;
-                let end = start + λ::USIZE;
-                let p_or_q = &self.u_or_v[start..end];
-                lagrange_table_p_or_q_r.eval(p_or_q)[0]
-            })
-            .collect();
+
+        let mut new_u_or_v_vec = Vec::<GenericArray<F, λ>>::new();
+        // iter over chunks of size λ
+        // and interpolate at x coordinate r
+        while let Some(polynomial) = u_or_v_iterator.next() {
+            let mut new_u_or_v = GenericArray::<F, λ>::generate(|_| F::ZERO);
+            new_u_or_v[0] = lagrange_table_p_or_q_r.eval(polynomial.borrow())[0];
+            for i in 1..λ::USIZE {
+                if let Some(polynomial) = u_or_v_iterator.next() {
+                    new_u_or_v[i] = lagrange_table_p_or_q_r.eval(polynomial.borrow())[0];
+                }
+            }
+            new_u_or_v_vec.push(new_u_or_v);
```

Comment:
Let's employ a similar approach to the one we went with on the last PR, that is creating a mutable generic array outside of the loop, and storing a mutable index, then having only a single while loop that over-writes the correct index of this array until it's full at which point it pushes it and creates a new one.